Using the dataset: FIFA 23 male players ( 5+ GB )

Have the csv file load in google drive and mounted it to run on colab.



In [1]:
from google.colab import drive
drive.mount('/content/drive')

# Path to the file in Google Drive
file_path = '/content/drive/MyDrive/male_players.csv'



Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
file_path = '/content/drive/MyDrive/male_players.csv'

# Test loading the first few lines
with open(file_path, 'r') as file:
    for _ in range(5):
        print(file.readline())


player_id,player_url,fifa_version,fifa_update,fifa_update_date,short_name,long_name,player_positions,overall,potential,value_eur,wage_eur,age,dob,height_cm,weight_kg,league_id,league_name,league_level,club_team_id,club_name,club_position,club_jersey_number,club_loaned_from,club_joined_date,club_contract_valid_until_year,nationality_id,nationality_name,nation_team_id,nation_position,nation_jersey_number,preferred_foot,weak_foot,skill_moves,international_reputation,work_rate,body_type,real_face,release_clause_eur,player_tags,player_traits,pace,shooting,passing,dribbling,defending,physic,attacking_crossing,attacking_finishing,attacking_heading_accuracy,attacking_short_passing,attacking_volleys,skill_dribbling,skill_curve,skill_fk_accuracy,skill_long_passing,skill_ball_control,movement_acceleration,movement_sprint_speed,movement_agility,movement_reactions,movement_balance,power_shot_power,power_jumping,power_stamina,power_strength,power_long_shots,mentality_aggression,mentality_interceptio

#Step 1: File reading methods

In this approach, the Pandas library is used to read the file in manageable chunks to handle large datasets efficiently.

Key Features:

Memory Efficiency: By specifying a chunksize, only a subset of the file is loaded into memory at any given time.

Flexible Processing: Each chunk can be processed independently, making it easier to handle large datasets without overloading the system.

In [3]:
import pandas as pd

# Define chunk size
chunk_size = 100_000

# Process file in chunks
for chunk in pd.read_csv(file_path, chunksize=chunk_size):
    print(chunk.info())  # Replace with your processing logic
    # Example: Calculate basic statistics
    print(chunk.describe())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(17), int64(46), object(47)
memory usage: 83.9+ MB
None
           player_id  fifa_version    fifa_update        overall  \
count  100000.000000      100000.0  100000.000000  100000.000000   
mean   235828.473540          23.0       6.770110      66.311180   
std     27368.342056           0.0       1.565639       6.856152   
min      1179.000000          23.0       4.000000      46.000000   
25%    219686.000000          23.0       5.000000      62.000000   
50%    239894.000000          23.0       7.000000      66.000000   
75%    258490.000000          23.0       8.000000      71.000000   
max    273129.000000          23.0       9.000000      91.000000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.947800e+04   99536.000000  100000.000000   
mean       71.169990  3.089508e+06    9567.312832   

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (85,86,87,88,89) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 3000000 to 3099999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(17), int64(46), object(47)
memory usage: 83.9+ MB
None
           player_id  fifa_version    fifa_update        overall  \
count  100000.000000      100000.0  100000.000000  100000.000000   
mean   222635.117800          20.0      36.240940      65.865910   
std     27851.310954           0.0       1.579251       6.895006   
min        41.000000          20.0      34.000000      46.000000   
25%    207499.000000          20.0      35.000000      61.000000   
50%    229658.000000          20.0      36.000000      66.000000   
75%    243397.000000          20.0      38.000000      70.000000   
max    256469.000000          20.0      39.000000      94.000000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.886500e+04   98904.000000  100000.000000   
mean       71.289930  2.364559e+06    9227.0

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 8000000 to 8099999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
           player_id   fifa_version    fifa_update        overall  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean   200432.173940      16.548660      26.927870      65.876060   
std     35702.125352       0.497629      27.408241       7.194152   
min         2.000000      16.000000       1.000000      44.000000   
25%    188000.000000      16.000000       2.000000      61.000000   
50%    208967.000000      17.000000       3.000000      66.000000   
75%    225364.000000      17.000000      57.000000      71.000000   
max    251838.000000      17.000000      58.000000      94.000000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.819700e+04   98607.000000  100000.000000   
mean       70.477830  1.857503e+06 

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 8600000 to 8699999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
           player_id  fifa_version    fifa_update        overall  \
count  100000.000000      100000.0  100000.000000  100000.000000   
mean   197461.955390          16.0      23.455650      65.043410   
std     36464.757498           0.0       1.698983       7.148931   
min         2.000000          16.0      20.000000      44.000000   
25%    185388.000000          16.0      22.000000      60.000000   
50%    206427.000000          16.0      23.000000      65.000000   
75%    222521.000000          16.0      25.000000      70.000000   
max    254664.000000          16.0      26.000000      94.000000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.816700e+04   98540.000000  100000.000000   
mean       69.763920  1.584443e+06   15210.4

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 8700000 to 8799999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
           player_id  fifa_version    fifa_update        overall  \
count  100000.000000      100000.0  100000.000000  100000.000000   
mean   196080.958090          16.0      17.536950      65.446030   
std     36698.217144           0.0       1.684833       6.952917   
min         2.000000          16.0      15.000000      40.000000   
25%    184287.000000          16.0      16.000000      61.000000   
50%    204988.500000          16.0      18.000000      65.000000   
75%    221283.000000          16.0      19.000000      70.000000   
max    254664.000000          16.0      20.000000      94.000000   

           potential     value_eur       wage_eur           age  \
count  100000.000000  9.810000e+04   98543.000000  100000.00000   
mean       69.972050  1.637660e+06   15817.693

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 8800000 to 8899999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
           player_id  fifa_version    fifa_update        overall  \
count  100000.000000      100000.0  100000.000000  100000.000000   
mean   195636.591220          16.0      11.696750      65.476140   
std     36855.093194           0.0       1.702868       7.119142   
min         2.000000          16.0       9.000000      44.000000   
25%    183897.000000          16.0      10.000000      61.000000   
50%    204626.000000          16.0      12.000000      66.000000   
75%    220883.000000          16.0      13.000000      70.000000   
max    254664.000000          16.0      15.000000      94.000000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.811700e+04   98574.000000  100000.000000   
mean       70.003890  1.701418e+06   16245.2

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 8900000 to 8999999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
           player_id  fifa_version    fifa_update        overall  \
count  100000.000000      100000.0  100000.000000  100000.000000   
mean   195759.103430          16.0       5.865210      65.367300   
std     36835.437321           0.0       1.731298       7.207174   
min         2.000000          16.0       3.000000      44.000000   
25%    183966.000000          16.0       4.000000      60.000000   
50%    204841.000000          16.0       6.000000      66.000000   
75%    220979.500000          16.0       7.000000      70.000000   
max    254664.000000          16.0       9.000000      94.000000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.811100e+04   98559.000000  100000.000000   
mean       70.113380  1.714464e+06   16261.5

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 9000000 to 9099999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
           player_id   fifa_version    fifa_update        overall  \
count  100000.000000  100000.000000  100000.000000  100000.000000   
mean   193773.610500      15.410410      34.729490      64.371510   
std     38286.010302       0.491911      27.512666       7.122546   
min         2.000000      15.000000       1.000000      40.000000   
25%    182859.750000      15.000000       2.000000      60.000000   
50%    203831.000000      15.000000      56.000000      64.000000   
75%    219667.000000      16.000000      58.000000      69.000000   
max    254664.000000      16.000000      59.000000      94.000000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.821100e+04   98547.000000  100000.000000   
mean       69.124050  1.286215e+06 

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 9100000 to 9199999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
           player_id  fifa_version   fifa_update        overall  \
count  100000.000000      100000.0  100000.00000  100000.000000   
mean   192314.414020          15.0      52.86620      63.756560   
std     39070.168712           0.0       1.79779       7.155976   
min         2.000000          15.0      50.00000      40.000000   
25%    181343.000000          15.0      51.00000      59.000000   
50%    202985.500000          15.0      53.00000      64.000000   
75%    216646.000000          15.0      54.00000      68.000000   
max    254664.000000          15.0      56.00000      93.000000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.820400e+04   98490.000000  100000.000000   
mean       68.501980  1.064606e+06   12955.498020    

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 9200000 to 9299999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
           player_id  fifa_version    fifa_update        overall  \
count  100000.000000      100000.0  100000.000000  100000.000000   
mean   191968.919490          15.0      46.785200      63.768630   
std     39323.182597           0.0       1.786458       7.163515   
min         2.000000          15.0      44.000000      40.000000   
25%    180714.000000          15.0      45.000000      59.000000   
50%    202800.000000          15.0      47.000000      64.000000   
75%    216444.000000          15.0      48.000000      68.000000   
max    254664.000000          15.0      50.000000      93.000000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.817100e+04   98490.000000  100000.000000   
mean       68.471850  1.068026e+06   13013.1

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 9300000 to 9399999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
           player_id  fifa_version    fifa_update        overall  \
count  100000.000000      100000.0  100000.000000  100000.000000   
mean   191793.283650          15.0      40.715410      63.765460   
std     39578.621262           0.0       1.776449       7.160088   
min         2.000000          15.0      38.000000      40.000000   
25%    180432.000000          15.0      39.000000      59.000000   
50%    202751.000000          15.0      41.000000      64.000000   
75%    216440.000000          15.0      42.000000      68.000000   
max    254664.000000          15.0      44.000000      93.000000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.816800e+04   98517.000000  100000.000000   
mean       68.479410  1.067237e+06   13021.6

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 9400000 to 9499999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
           player_id  fifa_version    fifa_update        overall  \
count  100000.000000      100000.0  100000.000000  100000.000000   
mean   191643.413030          15.0      34.630990      63.701760   
std     39522.159072           0.0       1.767742       7.193969   
min         2.000000          15.0      32.000000      40.000000   
25%    180283.000000          15.0      33.000000      59.000000   
50%    202691.000000          15.0      35.000000      64.000000   
75%    216295.000000          15.0      36.000000      68.000000   
max    254664.000000          15.0      38.000000      93.000000   

          potential     value_eur       wage_eur            age  \
count  100000.00000  9.816700e+04   98535.000000  100000.000000   
mean       68.43277  1.067273e+06   13022.3372

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 9500000 to 9599999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
           player_id  fifa_version    fifa_update        overall  \
count  100000.000000      100000.0  100000.000000  100000.000000   
mean   190872.786160          15.0      28.499420      63.852290   
std     39832.036175           0.0       1.771641       7.452572   
min         2.000000          15.0      25.000000      40.000000   
25%    179569.000000          15.0      27.000000      59.000000   
50%    202095.000000          15.0      29.000000      64.000000   
75%    215738.000000          15.0      30.000000      69.000000   
max    254664.000000          15.0      32.000000      93.000000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.820500e+04   98570.000000  100000.000000   
mean       68.563930  1.156597e+06   13926.2

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 9800000 to 9899999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
           player_id  fifa_version    fifa_update       overall  \
count  100000.000000      100000.0  100000.000000  100000.00000   
mean   189496.522530          15.0      10.090160      63.87619   
std     39729.173051           0.0       1.810112       7.06757   
min         2.000000          15.0       7.000000      40.00000   
25%    178042.000000          15.0       9.000000      59.00000   
50%    200827.000000          15.0      10.000000      64.00000   
75%    214524.000000          15.0      12.000000      68.00000   
max    254664.000000          15.0      13.000000      93.00000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.810800e+04   98541.000000  100000.000000   
mean       68.332220  1.067062e+06   13070.234725    

<ipython-input-3-ee29091037f7>:7: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv(file_path, chunksize=chunk_size):


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 9900000 to 9999999
Columns: 110 entries, player_id to player_face_url
dtypes: float64(18), int64(45), object(47)
memory usage: 83.9+ MB
None
          player_id  fifa_version    fifa_update        overall  \
count  100000.00000      100000.0  100000.000000  100000.000000   
mean   189251.53920          15.0       3.868130      63.810150   
std     39615.06826           0.0       1.857594       7.008566   
min         2.00000          15.0       1.000000      40.000000   
25%    178091.00000          15.0       2.000000      59.000000   
50%    200783.00000          15.0       4.000000      64.000000   
75%    214206.25000          15.0       5.000000      68.000000   
max    254664.00000          15.0       7.000000      93.000000   

           potential     value_eur       wage_eur            age  \
count  100000.000000  9.819700e+04   98647.000000  100000.000000   
mean       68.288720  1.054417e+06   12935.152615    

In [ ]:
from tqdm import tqdm

for chunk in tqdm(pd.read_csv(file_path, chunksize=100_000), desc="Reading file"):
    # Process each chunk
    pass


Reading file: 30it [01:19,  2.54s/it]/usr/local/lib/python3.10/dist-packages/tqdm/std.py:1181: DtypeWarning: Columns (85,86,87,88,89) have mixed types. Specify dtype option on import or set low_memory=False.
  for obj in iterable:
Reading file: 80it [03:37,  2.71s/it]/usr/local/lib/python3.10/dist-packages/tqdm/std.py:1181: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for obj in iterable:
Reading file: 86it [03:53,  2.79s/it]/usr/local/lib/python3.10/dist-packages/tqdm/std.py:1181: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for obj in iterable:
Reading file: 87it [03:56,  2.84s/it]/usr/local/lib/python3.10/dist-packages/tqdm/std.py:1181: DtypeWarning: Columns (108) have mixed types. Specify dtype option on import or set low_memory=False.
  for obj in iterable:
Reading file: 88it [03:59,  2.84s/it]/usr/local/lib/python3.10/dist-packages/tqdm/std.py:1181: DtypeWarning: 

Updated the dtype_mapping to reflect mixed and float data types.

In [4]:
dtype_mapping = {
    'cf': 'object',
    'club_contract_valid_until_year': 'float64',
    'club_jersey_number': 'float64',
    'club_loaned_from': 'object',
    'club_team_id': 'float64',
    'league_id': 'float64',
    'league_level': 'float64',
    'lf': 'object',
    'lw': 'object',
    'release_clause_eur': 'float64',
    'rf': 'object',
    'rw': 'object',
    'value_eur': 'float64',
    'wage_eur': 'float64'
}


Dask is a powerful library designed to handle large datasets that do not fit into memory. It reads files in chunks and processes them lazily, which makes it ideal for working with large CSV files like the 5GB+ file in this assignment.

Key Features of Dask:

Out-of-Memory Processing: Dask splits the file into manageable chunks, allowing processing on systems with limited memory.

Parallel Computing: It utilizes multiple CPU cores for faster data processing.


In [5]:
import dask.dataframe as dd

# Path to your file
file_path = '/content/drive/MyDrive/male_players.csv'

# Load the CSV file with specified dtypes
df = dd.read_csv(file_path, dtype=dtype_mapping)

# Preview the first few rows
print(df.head())


/usr/local/lib/python3.10/dist-packages/dask/dataframe/__init__.py:42: FutureWarning: 
Dask dataframe query planning is disabled because dask-expr is not installed.

You can install it with `pip install dask[dataframe]` or `conda install dask`.
This will raise in a future version.

  warnings.warn(msg, FutureWarning)


   player_id                                player_url  fifa_version  \
0     158023        /player/158023/lionel-messi/230009            23   
1     165153       /player/165153/karim-benzema/230009            23   
2     188545  /player/188545/robert-lewandowski/230009            23   
3     192985     /player/192985/kevin-de-bruyne/230009            23   
4     231747       /player/231747/kylian-mbappe/230009            23   

   fifa_update fifa_update_date      short_name  \
0            9       2023-01-13        L. Messi   
1            9       2023-01-13      K. Benzema   
2            9       2023-01-13  R. Lewandowski   
3            9       2023-01-13    K. De Bruyne   
4            9       2023-01-13       K. Mbappé   

                        long_name player_positions  overall  potential  ...  \
0  Lionel Andrés Messi Cuccittini               RW       91         91  ...   
1                   Karim Benzema           CF, ST       91         91  ...   
2              Robert L

#Features of PySpark for file reading:

Efficient Handling of Large Files:

PySpark can process datasets that are larger than your system's memory by distributing the workload across cores.

Parallel Processing:

PySpark divides the file into smaller partitions and processes them in parallel.


In [6]:
!pip install pyspark


In [22]:
from pyspark.sql import SparkSession

# Initialize SparkSession
spark = SparkSession.builder \
    .appName("LargeCSVReader") \
    .getOrCreate()

# Path to your large file
file_path = '/content/drive/MyDrive/male_players.csv'

# Read the CSV file
df = spark.read.csv(file_path, header=True, inferSchema=True)

# Show the first few rows
df.show(5)

# Print schema to understand column data types
df.printSchema()


+---------+--------------------+------------+-----------+----------------+--------------+--------------------+----------------+-------+---------+---------+--------+---+----------+---------+---------+---------+--------------+------------+------------+-------------------+-------------+------------------+----------------+----------------+------------------------------+--------------+----------------+--------------+---------------+--------------------+--------------+---------+-----------+------------------------+-------------+----------------+---------+------------------+--------------------+--------------------+----+--------+-------+---------+---------+------+------------------+-------------------+--------------------------+-----------------------+-----------------+---------------+-----------+-----------------+------------------+------------------+---------------------+---------------------+----------------+------------------+----------------+----------------+-------------+-------------+-

#Step 2: Data Validation

Data validation ensures that the dataset conforms to expected standards and formats. This is a crucial step in the pipeline to identify and correct errors, inconsistencies, and anomalies in the data.


###Objectives of Data Validation:

Column Validation:

1.   Check if all required columns are present.
2.   Validate column names and their data types.


Content Validation:


1.   Identify and handle missing or null values.
2.   Remove invalid or inconsistent data (e.g., negative ages or salaries).

3.   Check for duplicate rows.

Range and Constraint Validation:


1.   Verify numeric values fall within expected ranges.
2.   Validate categorical values against predefined sets.


In [40]:
for col in missing_columns:
    chunk[col] = None  # Add with default value (e.g., None or 0)




In [41]:
print("Updated columns:", chunk.columns)
print("DataFrame shape:", chunk.shape)


Updated columns: Index(['player_id', 'player_url', 'fifa_version', 'fifa_update',
       'fifa_update_date', 'short_name', 'long_name', 'player_positions',
       'overall', 'potential',
       ...
       'cdm', 'rdm', 'rwb', 'lb', 'lcb', 'cb', 'rcb', 'rb', 'gk',
       'player_face_url'],
      dtype='object', length=110)
DataFrame shape: (3590, 110)


In [42]:
required_columns = ['potential', 'value_eur', 'wage_eur', 'age', 'height_cm', 'weight_kg']
missing_columns = [col for col in required_columns if col not in chunk.columns]

if missing_columns:
    print(f"Missing columns: {missing_columns}")
else:
    print("All required columns are present.")


All required columns are present.


In [43]:
chunk.columns = chunk.columns.str.replace(r'[^\w]', '_').str.strip()
print("Validated column names:", chunk.columns.tolist())


Validated column names: ['player_id', 'player_url', 'fifa_version', 'fifa_update', 'fifa_update_date', 'short_name', 'long_name', 'player_positions', 'overall', 'potential', 'value_eur', 'wage_eur', 'age', 'dob', 'height_cm', 'weight_kg', 'league_id', 'league_name', 'league_level', 'club_team_id', 'club_name', 'club_position', 'club_jersey_number', 'club_loaned_from', 'club_joined_date', 'club_contract_valid_until_year', 'nationality_id', 'nationality_name', 'nation_team_id', 'nation_position', 'nation_jersey_number', 'preferred_foot', 'weak_foot', 'skill_moves', 'international_reputation', 'work_rate', 'body_type', 'real_face', 'release_clause_eur', 'player_tags', 'player_traits', 'pace', 'shooting', 'passing', 'dribbling', 'defending', 'physic', 'attacking_crossing', 'attacking_finishing', 'attacking_heading_accuracy', 'attacking_short_passing', 'attacking_volleys', 'skill_dribbling', 'skill_curve', 'skill_fk_accuracy', 'skill_long_passing', 'skill_ball_control', 'movement_accelerati

In [32]:
print(chunk.isnull().sum())  # Count missing values per column

# Option 1: Fill missing values
chunk['value_eur'] = chunk['value_eur'].fillna(0)

# Option 2: Drop rows with missing values
chunk = chunk.dropna()


potential    0
value_eur    0
wage_eur     0
age          0
height_cm    0
weight_kg    0
dtype: int64


In [33]:
expected_dtypes = {
    'potential': 'int64',
    'value_eur': 'float64',
    'wage_eur': 'float64',
    'age': 'int64',
    'height_cm': 'int64',
    'weight_kg': 'float64'
}

for col, dtype in expected_dtypes.items():
    if chunk[col].dtype != dtype:
        print(f"Converting {col} to {dtype}")
        chunk[col] = chunk[col].astype(dtype, errors='ignore')


Converting potential to int64
Converting value_eur to float64
Converting wage_eur to float64
Converting age to int64
Converting height_cm to int64
Converting weight_kg to float64


In [34]:
# Example: Validate ages
if (chunk['age'] < 16).any() or (chunk['age'] > 50).any():
    print("Invalid ages found.")
    chunk = chunk[(chunk['age'] >= 16) & (chunk['age'] <= 50)]

# Example: Validate wages
if (chunk['wage_eur'] < 0).any():
    print("Invalid wages found.")
    chunk = chunk[chunk['wage_eur'] >= 0]


In [35]:
duplicate_rows = chunk.duplicated()
print(f"Number of duplicate rows: {duplicate_rows.sum()}")

chunk = chunk.drop_duplicates()


Number of duplicate rows: 0


In [44]:
print(chunk.columns)  # Display all column names


Index(['player_id', 'player_url', 'fifa_version', 'fifa_update',
       'fifa_update_date', 'short_name', 'long_name', 'player_positions',
       'overall', 'potential',
       ...
       'cdm', 'rdm', 'rwb', 'lb', 'lcb', 'cb', 'rcb', 'rb', 'gk',
       'player_face_url'],
      dtype='object', length=110)


In [46]:
print(chunk['player_positions'].isnull().sum())  # Count null values
print(chunk['player_positions'].value_counts())  # Check actual values if any



0
player_positions
GK    541
Name: count, dtype: int64


In [47]:
print(chunk['player_positions'].unique())  # Verify the data


['GK']


In [48]:
valid_positions = ['GK', 'DEF', 'MID', 'FWD']

# Filter rows where valid positions are found
chunk = chunk[chunk['player_positions'].apply(lambda x: any(pos in x for pos in valid_positions))]


In [49]:
valid_positions = ['GK', 'DEF', 'MID', 'FWD']
if ~chunk['player_positions'].isin(valid_positions).all():
    print("Invalid player_positions found.")
    chunk = chunk[chunk['player_positions'].isin(valid_positions)]



In [50]:
print("Data validation completed!")
print("Number of rows after validation:", chunk.shape[0])
print("Number of columns:", chunk.shape[1])
print(chunk.head())


Data validation completed!
Number of rows after validation: 541
Number of columns: 110
          player_id                             player_url  fifa_version  \
10000002     190061     /player/190061/steven-deana/150001            15   
10000003     190179    /player/190179/david-cornell/150001            15   
10000008     191684         /player/191684/san-jung/150001            15   
10000009     191848  /player/191848/waleed-abdullah/150001            15   
10000013     193013       /player/193013/james-shea/150001            15   

          fifa_update fifa_update_date   short_name            long_name  \
10000002            1       2014-08-29     S. Deana         Steven Deana   
10000003            1       2014-08-29   D. Cornell        David Cornell   
10000008            1       2014-08-29     Jung San                정산 郑山   
10000009            1       2014-08-29  W. Abdullah  Waleed Abdullah Ali   
10000013            1       2014-08-29      J. Shea           James Shea   


In [51]:
chunk.to_csv('validated_chunk.csv', mode='a', index=False, header=False)


#Step 3:Schema Definition Using YAML


Create a YAML File
Define the schema for your dataset in a schema.yaml file.

####Specify:

Column names.

Expected data types.

Constraints like required columns or allowed ranges.


In [52]:
!pip install pyyaml


In [53]:
import yaml

# Load the schema from the YAML file
with open('/content/drive/MyDrive/Colab Notebooks/schema.yaml', 'r') as file:  # Provide the correct file path
    schema = yaml.safe_load(file)

print("Schema loaded successfully:")
print(schema)


Schema loaded successfully:
{'schema': {'columns': [{'potential': {'type': 'integer', 'description': 'Player potential rating'}}, {'value_eur': {'type': 'float', 'description': 'Player value in euros'}}, {'wage_eur': {'type': 'float', 'description': 'Player wage in euros'}}, {'age': {'type': 'integer', 'description': 'Player age'}}, {'height_cm': {'type': 'integer', 'description': 'Player height in centimeters'}}, {'weight_kg': {'type': 'float', 'description': 'Player weight in kilograms'}}, {'player_positions': {'type': 'string', 'description': 'Player position on the field'}}, {'club_name': {'type': 'string', 'description': 'The club the player belongs to'}}, {'nationality_name': {'type': 'string', 'description': 'Player nationality'}}], 'validation_rules': ['Age should be between 16 and 50.', 'Wage must be non-negative.', 'Player value must be positive.']}}


In [54]:
# Extract expected column names from the schema
expected_columns = [list(col.keys())[0] for col in schema['schema']['columns']]

# Check for missing columns
missing_columns = [col for col in expected_columns if col not in chunk.columns]

if missing_columns:
    print(f"Missing columns: {missing_columns}")
else:
    print("All required columns are present.")



All required columns are present.


In [55]:
# Define a mapping between YAML data types and Python/pandas types
dtype_mapping = {
    'integer': 'int64',
    'float': 'float64',
    'string': 'object'
}

# Validate data types
for column in schema['schema']['columns']:
    col_name = list(column.keys())[0]  # Column name
    expected_type = column[col_name]['type']
    actual_type = str(chunk[col_name].dtype)

    # Check if the data type matches
    if actual_type != dtype_mapping[expected_type]:
        print(f"Column '{col_name}' has invalid type. Expected {dtype_mapping[expected_type]}, found {actual_type}.")
    else:
        print(f"Column '{col_name}' has a valid type.")


Column 'potential' has a valid type.
Column 'value_eur' has a valid type.
Column 'wage_eur' has a valid type.
Column 'age' has a valid type.
Column 'height_cm' has a valid type.
Column 'weight_kg' has invalid type. Expected float64, found int64.
Column 'player_positions' has a valid type.
Column 'club_name' has a valid type.
Column 'nationality_name' has a valid type.


In [56]:
# Example: Validate ranges
if 'age' in chunk.columns:
    if (chunk['age'] < 16).any() or (chunk['age'] > 50).any():
        print("Invalid age values found.")
        chunk = chunk[(chunk['age'] >= 16) & (chunk['age'] <= 50)]

# Example: Non-negative validation for wages
if 'wage_eur' in chunk.columns:
    if (chunk['wage_eur'] < 0).any():
        print("Invalid wage values found.")
        chunk = chunk[chunk['wage_eur'] >= 0]


In [57]:
# Save the validated chunk
chunk.to_csv('/content/drive/MyDrive/validated_chunk.csv', mode='a', index=False)
print("Validated data saved successfully.")


Validated data saved successfully.


#Step 4: Writing and Summarizing the Validated Dataset

File Writing:

The file was compressed into a .gz format to minimize storage requirements and enhance data handling efficiency.


In [58]:
# Write the DataFrame to a pipe-separated text file compressed in gzip format
output_file = '/content/drive/MyDrive/validated_data.psv.gz'

chunk.to_csv(output_file, sep='|', index=False, compression='gzip')
print(f"File written successfully to {output_file}")


File written successfully to /content/drive/MyDrive/validated_data.psv.gz


In [59]:
import os

# Calculate total number of rows and columns
total_rows, total_columns = chunk.shape

# Calculate file size
file_size = os.path.getsize(output_file) / (1024**2)  # Convert size to MB

# Print the summary
print(f"File Summary:")
print(f"Total Rows: {total_rows}")
print(f"Total Columns: {total_columns}")
print(f"File Size: {file_size:.2f} MB")


File Summary:
Total Rows: 541
Total Columns: 110
File Size: 0.06 MB


File Summary:

Total Rows: The dataset contains 541 rows, indicating the number of observations retained after validation.

Total Columns: The dataset includes 110 columns, covering all relevant variables required for analysis.

File Size: The compressed file size is approximately 0.06 MB, highlighting the effectiveness of the GZ compression.